cmake/ThirdParty.cmake

```diff
@@ -180,6 +180,30 @@ include_directories(SYSTEM
   ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/cli11-prefix/include
 )
 
+#gtest
+if(WITH_TESTS)
+  ExternalProject_Add(googletest
+    GIT_REPOSITORY    https://github.com/google/googletest.git
+    GIT_TAG           v1.14.0
+    GIT_SHALLOW       ON
+    SOURCE_DIR        "${CMAKE_CURRENT_BINARY_DIR}/googletest-src"
+    BINARY_DIR        "${CMAKE_CURRENT_BINARY_DIR}/googletest-build"
+    INSTALL_COMMAND   ""
+    TEST_COMMAND      ""
+    LOG_DOWNLOAD      ON
+    LOG_BUILD         ON
+  )
+
+  include_directories(SYSTEM
+    "${CMAKE_CURRENT_BINARY_DIR}/googletest-src/googletest/include"
+    "${CMAKE_CURRENT_BINARY_DIR}/googletest-src/googlemock/include"
+  )
+
+  find_package(GTest REQUIRED)
+
+  enable_testing()
+endif()
```

Comment:
Thanks for adding google test!

Comments:

* I was planning to switch dependencies from submodules to downloading them. Submodules approach does not allow to build s-e from source tarball, generated by github for release. Since here you also are downloading gtest, I did what I planned and pushed the change: c09f02382a0afc1edffee8d8eb2f873c44110c34

* We can't use find_package here. find_package works and configure time (when cmake is running). However, ExternalProject_Add works at build time (when make is running after cmake finishes). So find_package won't find our downloaded gtest. I guess on your computer it finds gtest installed system-wide, and on CI it fails.

* Instead of using find_package, please use the same pattern as for all other dependencies. We need to build gtest, install into prefix dir, and use installed headers and libs from that prefix dir. We also need to forward all cmake arguments to support cross-compilation. Something like this:

```cmake
# googletest
ExternalProject_Add(googletest_lib
  GIT_REPOSITORY "https://github.com/google/googletest.git"
  GIT_TAG "v1.14.0"
  GIT_SHALLOW ON
  SOURCE_DIR ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/googletest-src
  BINARY_DIR ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/googletest-build
  PREFIX ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/googletest-prefix
  CMAKE_ARGS
    -DCMAKE_INSTALL_PREFIX=<INSTALL_DIR>
    -DCMAKE_TOOLCHAIN_FILE=${CMAKE_TOOLCHAIN_FILE}
    -DCMAKE_C_COMPILER_TARGET=${CMAKE_C_COMPILER_TARGET}
    -DCMAKE_CXX_COMPILER_TARGET=${CMAKE_CXX_COMPILER_TARGET}
    -DCMAKE_C_COMPILER=${CMAKE_C_COMPILER}
    -DCMAKE_CXX_COMPILER=${CMAKE_CXX_COMPILER}
    -DCMAKE_AR=${CMAKE_AR}
    -DCMAKE_RANLIB=${CMAKE_RANLIB}
    -DCMAKE_STRIP=${CMAKE_STRIP}
    -DCMAKE_POSITION_INDEPENDENT_CODE=ON
    -DKISSFFT_PKGCONFIG=OFF
    -DKISSFFT_STATIC=ON
    -DKISSFFT_TEST=OFF
    -DKISSFFT_TOOLS=OFF
    -DKISSFFT_USE_ALLOCA=OFF
  LOG_DOWNLOAD YES
  LOG_CONFIGURE YES
  LOG_BUILD YES
  LOG_INSTALL YES
)
include_directories(SYSTEM
  ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/googletest-prefix/include
)
link_libraries(
  ${CMAKE_CURRENT_BINARY_DIR}/3rdparty/googletest-prefix/lib/${LIBPREFIX}gtest${LIBSUFFIX}
)
```

* Please also add `googletest_lib` to ALL_DEPENDENCIES at the bottom of the file.

* Please move enable_testing() call to root cmake file.

Reply:
Yes, currently the CI fails and I wanted to have another look into that. I currently use `find_library` because otherwise I was not able to use `gtest_discover_tests()`. I have to check how to use that function without using find_package.